hugr-passes/src/dataflow/test.rs

```diff
@@ -0,0 +1,498 @@
+    machine.run(TestContext(Arc::new(&hugr)));
+
+    let o_r1 = machine.read_out_wire(o_w1).unwrap();
+    assert_eq!(o_r1, pv_true());
+    let o_r2 = machine.read_out_wire(o_w2).unwrap();
+    assert_eq!(o_r2, pv_false());
+    assert_eq!(
+        Some(TailLoopTermination::BreaksAndContinues),
+        machine.tail_loop_terminates(&hugr, tail_loop.node())
+    );
+    assert_eq!(machine.tail_loop_terminates(&hugr, hugr.root()), None);
+}
+
+#[test]
+fn conditional() {
+    let variants = vec![type_row![], type_row![], type_row![BOOL_T]];
+    let cond_t = Type::new_sum(variants.clone());
+    let mut builder = DFGBuilder::new(Signature::new(cond_t, type_row![])).unwrap();
+    let [arg_w] = builder.input_wires_arr();
+
+    let true_w = builder.add_load_value(Value::true_val());
+    let false_w = builder.add_load_value(Value::false_val());
+
+    let mut cond_builder = builder
+        .conditional_builder(
+            (variants, arg_w),
+            [(BOOL_T, true_w)],
+            type_row!(BOOL_T, BOOL_T),
+        )
+        .unwrap();
+    // will be unreachable
+    let case1_b = cond_builder.case_builder(0).unwrap();
+    let case1 = case1_b.finish_with_outputs([false_w, false_w]).unwrap();
+
+    let case2_b = cond_builder.case_builder(1).unwrap();
+    let [c2a] = case2_b.input_wires_arr();
+    let case2 = case2_b.finish_with_outputs([false_w, c2a]).unwrap();
+
+    let case3_b = cond_builder.case_builder(2).unwrap();
+    let [c3_1, _c3_2] = case3_b.input_wires_arr();
+    let case3 = case3_b.finish_with_outputs([c3_1, false_w]).unwrap();
+
+    let cond = cond_builder.finish_sub_container().unwrap();
+
+    let [cond_o1, cond_o2] = cond.outputs_arr();
+
+    let hugr = builder.finish_hugr(&EMPTY_REG).unwrap();
+
+    let mut machine = Machine::default();
+    let arg_pv = PartialValue::new_variant(1, []).join(PartialValue::new_variant(
+        2,
+        [PartialValue::new_variant(0, [])],
+    ));
+    machine.propolutate_out_wires([(arg_w, arg_pv)]);
+    machine.run(TestContext(Arc::new(&hugr)));
+
+    let cond_r1 = machine
+        .read_out_wire(cond_o1)
+        .unwrap()
+        .try_into_wire_value(&hugr, cond_o1)
+        .unwrap();
+    assert_eq!(cond_r1, Value::false_val());
+    assert!(machine
+        .read_out_wire(cond_o2)
+        .unwrap()
+        .try_into_wire_value(&hugr, cond_o2)
+        .is_err());
+
+    assert_eq!(machine.case_reachable(&hugr, case1.node()), Some(false)); // arg_pv is variant 1 or 2 only
+    assert_eq!(machine.case_reachable(&hugr, case2.node()), Some(true));
+    assert_eq!(machine.case_reachable(&hugr, case3.node()), Some(true));
+    assert_eq!(machine.case_reachable(&hugr, cond.node()), None);
+}
+
+// Tuple of
+//   1. Hugr being a function on bools: (x, y) => (x XOR y, x AND y)
+//   2. Input node of entry block
+// Result readable from root node outputs
+// Inputs should be placed onto out-wires of the Node (2.)
+#[fixture]
+fn xor_and_cfg() -> (Hugr, Node) {
```

Comment:
I am realizing this is getting to be a seriously complex CFG. Originally I had the goal that we should test on a CFG that does something that can't easily be done in a Conditional or TailLoop. Making this also cover the need for reachability on CFG basic blocks has made it (slightly) more complex still, so could move that into a separate test (e.g. on a simple diamond CFG), although the improvement here would be slight. Or I could just make this test a CFG that's equivalent to a conditional, say...